### PR TITLE
[ROCm] Forward device function name from gpu prefix to cuda prefix

### PR DESCRIPTION
**Pull request by jerryyin**
This PR is a follow-up to the original PR #28343. The reviewer requested to break down the original large PR to a series of smaller ones. According to the plan here, this PR is the 3rd one in the whole series. It is dependent on the infrastructure provided from PR #28564

@chsigg @whchung

### CODE REVIEW

File: tensorflow/core/util/gpu_device_functions.h
Comment (chsigg) [inline tensorflow/core/util/gpu_device_functions.h]: You do start calling the Gpu variant here and in a number of places below. The PR is still manageable, so let's take this shortcut.

File: tensorflow/core/util/gpu_device_functions.h
Comment (chsigg) [inline tensorflow/core/util/gpu_device_functions.h]: Is it valid to add the alias before the template specializations?

File: tensorflow/core/util/gpu_kernel_helper.h
Comment (chsigg) [inline tensorflow/core/util/gpu_kernel_helper.h]: This change seems unrelated. Is it inherited from a different PR that is supposed to go in first?

File: tensorflow/core/util/gpu_device_functions.h
Comment (jerryyin) [reply tensorflow/core/util/gpu_device_functions.h]: Sounds good, thanks

File: tensorflow/core/util/gpu_kernel_helper.h
Comment (jerryyin) [reply tensorflow/core/util/gpu_kernel_helper.h]: <backgroud> I don't know why it ended up in this location in our develop-upstream branch, because the definition started in the gpu_launch_config.h. It might be handled differently in our weekly merge task.

I'd assume you prefer it in non-macro format as well. Will update in next commit.

File: tensorflow/core/util/gpu_device_functions.h
Comment (jerryyin) [reply tensorflow/core/util/gpu_device_functions.h]: I used to have this concern as well. Looking at what the macro does, it does a text replacement to a perfect forwarding signature of the new function name. So the question ended up as whether the perfect forwarding declaration is location relevant. 

I'd say it is not because when compiler does the actual forwarding, it needs to match up the types and number of arguments. The process of matching up the correct function signature should be as the same as if the compiler resolves any other template functions.

Comment (chsigg) [reply tensorflow/core/util/gpu_device_functions.h]: I think you are correct, the compiler has seen all template specializations when it instantiates the aliasing template.

File: tensorflow/core/util/gpu_kernel_helper.h
Comment (jerryyin) [reply tensorflow/core/util/gpu_kernel_helper.h]: #28568 commit a7137e1 addressed the review by moving GetGpuStream() from gpu_launch_config.h to gpu_kernel_helper.h

Will remove the wrong `GetGPUStream()` macro.

File: tensorflow/core/util/gpu_device_functions.h
Comment (chsigg) [inline tensorflow/core/util/gpu_device_functions.h]: Wasn't this removed at HEAD?

File: tensorflow/core/util/gpu_kernel_helper.h
Comment (chsigg) [reply tensorflow/core/util/gpu_kernel_helper.h]: Let's just leave it for now.

File: tensorflow/core/util/gpu_device_functions.h
Comment (jerryyin) [reply tensorflow/core/util/gpu_device_functions.h]: That's right. Done

**Review by whchung — CHANGES_REQUESTED** [review]
@jerryyin , in other places of `CREATE_CUDA_DEVICE_FUNCTION_ALIAS()` the macro is used at the bottom of all variations of function templates or functions, but not for `GpuLdg`. please help change it.

**Comment by whchung** [comment]
nit picking: remove the empty line (line 111), between `GpuGridRangeX` and `CREATE_CUDA_DEVICE_FUNCTION_ALIAS`. github UI somehow prevents me commenting in-place.

**Comment by jerryyin** [comment]
@whchung Sounds good. I'll do a force push to avoid additional commit. For your convenience, the diff is below:

> @@ -108,7 +108,6 @@ __device__ detail::GpuGridRange<T> GpuGridRangeX(T count) {
   return detail::GpuGridRange<T>(blockIdx.x * blockDim.x + threadIdx.x,
                                   gridDim.x * blockDim.x, count);
 }
 \-
CREATE_CUDA_DEVICE_FUNCTION_ALIAS(GpuGridRangeX, CudaGridRangeX);
// Helper to visit indices in the range 0 <= i < count using the y-coordinate.
   @@ -467,7 +466,6 @@ __host__ __device__ T GpuLdg(const T* address) {
    return *address;
 #endif
 }
\-CREATE_CUDA_DEVICE_FUNCTION_ALIAS(GpuLdg, CudaLdg);
__host__ __device__ inline bool GpuLdg(const bool* address) {
   return GpuLdg(reinterpret_cast<const char*>(address)) != 0;
@@ -492,6 +490,7 @@ __host__ __device__ inline std::complex<double> GpuLdg(
   return *address;
 #endif
 }
+CREATE_CUDA_DEVICE_FUNCTION_ALIAS(GpuLdg, CudaLdg);

**Review by whchung — DISMISSED** [review]

**Comment by jerryyin** [comment]
@chsigg There was a point when I get your approval, but lost that due to the need to rebase to fix conflicts. Would you mind taking a second look at this PR, and re-approve is possible? I'm asking because this PR will starts blocking others in ROCm team if not merged.

File: tensorflow/core/util/gpu_device_functions.h
Comment (chsigg) [inline tensorflow/core/util/gpu_device_functions.h]: The two are the same, could you just leave the comment but remove the #if/else?

File: tensorflow/core/util/gpu_device_functions.h
Comment (chsigg) [inline tensorflow/core/util/gpu_device_functions.h]: Do you want CudaGridRange conditionally, or could you just remove the #if/else?

File: tensorflow/core/util/gpu_device_functions.h
Comment (jerryyin) [reply tensorflow/core/util/gpu_device_functions.h]: Make sense, will do.

File: tensorflow/core/util/gpu_device_functions.h
Comment (jerryyin) [reply tensorflow/core/util/gpu_device_functions.h]: I want it conditionally, as in most cases of `CREATE_CUDA_*`. Right now the type is also used in `segment_reduction_ops_gpu.cu.cc` and `trdiagonal_solve_op_gpu.cu.cc` kernels. Having it conditionally available in cuda will make those code compile. 

Sometime in the future when we send those (ROCm available) kernels up to review we will update the symbol to be using GpuGridRange and get rid of the conditional variable here.

File: tensorflow/core/util/gpu_device_functions.h
Comment (jerryyin) [reply tensorflow/core/util/gpu_device_functions.h]: Taking a step back, I feel like keeping it and get the ROCm side of the `kCudaWarpAll` to be renamed to `kGpuWarpAll`, just to be consistent to the aim of creating alias. Since the aim was to make the `Gpu` in variable naming so that it works for both ROCm and Cuda. 

This is so that when dependent kernels that uses `kCudaWarpAll`, as well as line 133 will be renamed to kGpuWarpAll as well. How does that sound?

Comment (chsigg) [reply tensorflow/core/util/gpu_device_functions.h]: We have been providing both symbols during a rename. If you just rename the ROCm side, we won't be able to change usages one by one.

File: tensorflow/core/util/gpu_device_functions.h
Comment (chsigg) [reply tensorflow/core/util/gpu_device_functions.h]: I would rather be able to change call sites independently of adding ROCm support to the kernels. The two things are unrelated and should not merged into single PRs.

File: tensorflow/core/util/gpu_device_functions.h
Comment (jerryyin) [reply tensorflow/core/util/gpu_device_functions.h]: Sounds good. Will remove #if/else and make both kCudaWarpAll and kGpuWarpAll available to be consistent with the reset of PR.

File: tensorflow/core/util/gpu_device_functions.h
Comment (jerryyin) [reply tensorflow/core/util/gpu_device_functions.h]: Good clarification.

@whchung FYI. Since in our codebase, all call-sites are already renamed. If we start with ROCm support of individual kernels, we will end up doing two things together. Therefore it looks like the right sequence is to:

- File individual PRs to each symbols from `cuda` to `gpu`, one symbol in a PR at a time.
- File individual PRs to enable one kernel as a time.

Comment (deven-amd) [reply tensorflow/core/util/gpu_device_functions.h]: @chsigg, @jerryyin 

I want to make sure we are on the same page here.

Once this PR is merged, we will have the 
1. Gpu* names visiblle in both ROCm and CUDA builds
2. Cuda* names visible only in CUDA builds

Once this PR is merged, we will also start filing PRs for ops that reference these names (in the same 1 PR/op pattern we have established so far). As part of the changes for such a PR, our intent is to update the reference(s) to the Cuda* name(s) to Gpu* name(s).  This implies that the renaming will be spread out across multiple PRs.

If it is preferable to have one PR, that does the Cuda* to Gpu* renaming for all references in a single shot (@chsigg, this is what my interpretation of your comment was...please correct if wrong), then we need to file that PR before we file any other PRs to upstream ROCm support. Remember that Cuda* names are not visible in ROCm builds!

Comment (chsigg) [reply tensorflow/core/util/gpu_device_functions.h]: > Remember that Cuda* names are not visible in ROCm builds!
Right, I forgot that. So the #ifndef here is consistent with the rest. 

It will be easier if we change the call sites from Cuda* to Gpu* on our side. We have dependencies that you don't see and infrastructure to help making the change. It will be a series of CLs.

**Review by chsigg — APPROVED** [review]

**Pull request merged**
